lib/lambda_ethereum_consensus/beacon/pending_blocks.ex

```diff
@@ -114,65 +112,45 @@ defmodule LambdaEthereumConsensus.Beacon.PendingBlocks do
     end)
   end
 
-  @empty_mapset MapSet.new()
-
-  @impl true
-  def handle_info(:download_blocks, %{blocks_to_download: to_download} = state)
-      when to_download == @empty_mapset do
-    schedule_blocks_download()
-    {:noreply, state}
-  end
-
   @impl true
   def handle_info(:download_blocks, state) do
-    blocks_in_store = state.blocks_to_download |> MapSet.filter(&ForkChoice.has_block?/1)
+    blocks_to_download = state |> Map.filter(fn {_, {_, s}} -> s == :download end) |> Map.keys()
 
     downloaded_blocks =
-      state.blocks_to_download
-      |> MapSet.difference(blocks_in_store)
-      |> Enum.take(16)
-      |> BlockDownloader.request_blocks_by_root()
-      |> case do
-        {:ok, signed_blocks} ->
-          signed_blocks
-
-        {:error, reason} ->
-          Logger.debug("Block download failed: '#{reason}'")
-          []
+      if Enum.empty?(blocks_to_download) do
+        []
+      else
+        blocks_to_download
+        |> Enum.take(16)
+        |> BlockDownloader.request_blocks_by_root()
+        |> case do
+          {:ok, signed_blocks} ->
+            signed_blocks
+
+          {:error, reason} ->
+            Logger.debug("Block download failed: '#{reason}'")
+            []
+        end
       end
 
-    for signed_block <- downloaded_blocks do
-      add_block(signed_block)
-    end
-
-    roots_to_remove =
+    new_state =
       downloaded_blocks
-      |> Enum.map(&Ssz.hash_tree_root!(&1.message))
-      |> MapSet.new()
-      |> MapSet.union(blocks_in_store)
+      |> Enum.reduce(state, fn signed_block, state ->
+        block_root = Ssz.hash_tree_root!(signed_block.message)
+        state |> Map.put(block_root, {signed_block, :pending})
```

Comment:
Maybe both here and when casting `:add_block` we should call a single function that does the ssz hashing + adding the block to the state as pending? Also in that function we could validate that the block is not already present in the state, as in some cases we might get the block from two different sources (e.g. through gossip and manual download) and we probably don't want to replace an existing entry.